lint/rule.go

```diff
@@ -14,10 +14,172 @@
 
 package lint
 
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
+
 // Rule defines a lint rule that checks Google Protobuf APIs.
-type Rule interface {
-	// Info returns metadata about a rule.
-	Info() RuleInfo
-	// Lint performs the linting process.
-	Lint(Request) ([]Problem, error)
+type Rule struct {
+	// Name contains the rule's name.
+	Name RuleName
+
+	// Description is a short description of the rule,
+	// used in documentation and elsewhere.
+	Description string
+
+	// URI is the address where the guideline is documented.
+	// This should be displayed to give API designers more information about
+	// "how to do this right".
+	URI string
+
+	// LintFile is called for files as a whole.
+	LintFile func(*desc.FileDescriptor) []Problem
```

Comment:
Why the removal of `error`? The inability for a rule to express that they failed to lint a proto seems like a limitation. Tricorder also has hooks to report linting failures, which are useful for debugging.